velox/exec/tests/RowContainerTest.cpp

```diff
@@ -187,6 +190,60 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
     testRowNumbers();
   }
 
+  void testRowNumbersSpecialCases(
+      RowContainer& container,
+      const std::vector<char*>& rows,
+      int column,
+      const VectorPtr& expected) {
+    auto size = rows.size();
+    auto testRepeatedRowNumber = [&]() {
+      // Copy the first row to all positions in the result vector
+      std::vector<vector_size_t> rowNumbers;
+      rowNumbers.resize(size);
+      for (int i = 0; i < size; i++) {
+        rowNumbers[i] = 0;
+      }
+      folly::Range<const vector_size_t*> rowNumbersRange =
+          folly::Range(rowNumbers.data(), size);
+
+      auto result = BaseVector::create(expected->type(), size, pool_.get());
+      container.extractColumn(rows.data(), rowNumbersRange, column, 0, result);
+
+      EXPECT_EQ(size, result->size());
+      for (vector_size_t i = 0; i < size; ++i) {
+        EXPECT_TRUE(result->equalValueAt(expected.get(), i, 0))
+            << "at " << i << ": expected " << expected->toString(0) << ", got "
+            << result->toString(i);
+      }
+    };
+
+    auto testNegativeRowNumber = [&]() {
+      // -ve row numbers result in nullptr being copied to that position.
+      std::vector<vector_size_t> rowNumbers;
+      rowNumbers.resize(size);
+      for (int i = 0; i < size; i++) {
+        rowNumbers[i] = -i;
+      }
+      folly::Range<const vector_size_t*> rowNumbersRange =
+          folly::Range(rowNumbers.data(), size);
+
+      auto result = BaseVector::create(expected->type(), size, pool_.get());
+      container.extractColumn(rows.data(), rowNumbersRange, column, 0, result);
+
+      EXPECT_EQ(size, result->size());
+      EXPECT_TRUE(result->equalValueAt(expected.get(), 0, 0))
+          << "at 0 expected " << expected->toString(0) << ", got "
+          << result->toString(0);
+      for (vector_size_t i = 1; i < size; ++i) {
+        EXPECT_TRUE(result->isNullAt(i))
+            << "at " << i << "expected null, got " << result->toString(i);
+      }
+    };
+
+    testRepeatedRowNumber();
+    testNegativeRowNumber();
```

Comment:
Would it make sense to test a mix of valid offsets and negative offsets? Would it also make sense to test repeated offsets that are not all the same?